### PR TITLE
UI: Glimmerize search select component

### PR DESCRIPTION
**Pull request by hellobontempo**
This glimmerizes the <SearchSelect> component, along with cleans up some default stylistic behavior now that we are using the component more and more outside of `<FormField>`

- Changes the component's default label class from being a section header `"title is-4"` to the smaller input label size `"is-label"`.
- Removes `subLabel` - this arg was redundant and used just to render a smaller label. Consulted design in the one case where both existed and it was deemed unnecessary.
- Adds `isSectionHeader` as an option for search select's rendered from `FormField`
- Removes one-off args passed to `<SearchSelect>` to cleanup the component's params and instead handle specific customization in the parent 
- Makes `@renderInfoTooltip` a function, so the parent has more control over customizing why a tooltip renders and what the text is
- Adds more explicit test coverage for `SearchSelect`

### CODE REVIEW

File: ui/app/templates/components/mfa/mfa-login-enforcement-form.hbs
Comment (hellobontempo) [inline ui/app/templates/components/mfa/mfa-login-enforcement-form.hbs:107]: curious about using `@models={{array this.selectedTargetType}}` instead - which I _think_ would clear up the `.js` file for this form?

File: ui/app/models/database/role.js
Comment (hellobontempo) [inline ui/app/models/database/role.js:15]: I think the intent here may have been to only render a smaller label instead of a section header. But `FormField` generates a `labelString` if one isn't passed in, so two labels were rendering. 

Before 
Screen Shot 2022-09-26 at 12 20 11 PM

Now 
Screen Shot 2022-09-26 at 12 20 26 PM

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hellobontempo) [inline ui/app/templates/components/get-credentials-card.hbs:17]: I think `@inputValue` was copied over accidentally as it's unnecessary here. It's used for forms to pre-fill selected items, which happens on edit

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hellobontempo) [inline ui/app/templates/components/get-credentials-card.hbs:15]: having `shouldUseFallback` as an arg _and_ tracked variable set by the component was confusing (it was set to `true` if the query returned a `403`)

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hellobontempo) [inline ui/app/templates/components/get-credentials-card.hbs:26]: While I was in here cleaned this up - but if it feels outside the scope of this PR can remove!

File: ui/app/components/get-credentials-card.js
Comment (hellobontempo) [inline ui/app/components/get-credentials-card.js:16]: > Only works if shouldUseFallback is true.

This wasn't actually true, there was just a typo in the `SearchSelect` template. Updated everywhere in the code so now `placeHolder` is `placeholder` for consistency

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [inline ui/lib/core/addon/components/search-select.js:69]: Kept the default `false` per the original component. 

@ivana-mcconnell - we discussed last UI sync to change this so that SearchSelect _does not_ allowing "creating" new items by default, right? As it's confusing since you're not actually creating the thing, just associating the name of the thing. Should I go ahead and make that change?

File: ui/app/templates/components/mfa/mfa-login-enforcement-form.hbs
Comment (hellobontempo) [inline ui/app/templates/components/mfa/mfa-login-enforcement-form.hbs:109]: Now that `selectedOptions` is a tracked property, we don't want to set it directly so created a new arg - ideally opening up the way to refactor for a parent to manage the selected items

File: ui/lib/core/addon/components/search-select.js
Comment (ivana-mcconnell) [reply ui/lib/core/addon/components/search-select.js:69]: That's right— remove the ability to create new things here, since we can't (and won't) show the full database connection form here.

File: ui/app/models/database/role.js
Comment (Monkeychip) [reply ui/app/models/database/role.js:15]: The "Now" looks correct, the only difference from the designs below is that we have Role name first in the page view and not in the designs 🤷‍♀️

image

File: ui/app/templates/components/get-credentials-card.hbs
Comment (Monkeychip) [reply ui/app/templates/components/get-credentials-card.hbs:26]: Thank you for the cleanup. The only reason to remove to another PR is if you feel this would need to be backported? If not, I'm fine keeping it in.

File: ui/lib/core/addon/components/search-select-with-modal.js
Comment (Monkeychip) [inline ui/lib/core/addon/components/search-select-with-modal.js:37]: Is there somewhere where we accepted a comma-seperated string? I know in input fields we've removed tooltips that said we did (the cli allows it but we don't in those fields). Just want to make sure it's not legacy functionality we just don't use anymore.

File: ui/app/models/database/role.js
Comment (hellobontempo) [reply ui/app/models/database/role.js:15]: Thanks for the screenshot of designs! It also looks like they're not separate sections in the designs (but in the UI they're separated)

File: ui/lib/core/addon/components/search-select-with-modal.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select-with-modal.js:37]: Yeah, I poked around before removing this documentation. The biggest indicator for me was the search select file had no way of handling a string. If a comma separated string was passed in, it didn't work because it couldn't map over it. Thanks for following up!

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hellobontempo) [reply ui/app/templates/components/get-credentials-card.hbs:26]: I don't feel like it needs to be backported? That's a good point about button vs input. I'll confirm with design!

File: ui/app/models/database/role.js
Comment (zofskeez) [reply ui/app/models/database/role.js:15]: It seems to me if we are creating a role then _Role name_ should be the first field.

File: ui/lib/core/addon/components/search-select.hbs
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.hbs:19]: I wonder if this would be better as a `yield` in the future? Maybe not if it's typically the same component like an input passed in but just a thought.

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:63]: Does this need to be here if we're not setting an initial value?

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:33]: It looks like the arg name should be `parentManageSelected`

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:91]: I think unless we need to check for strict equality of `false` these getters are unnecessary.

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:94]: I never think to use `array.some` and always reach for `find`. Nice!

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:136]: :tada:

File: ui/lib/core/addon/components/search-select.js
Comment (zofskeez) [inline ui/lib/core/addon/components/search-select.js:150]: It looks like this could be combined into one if statement and you could get rid of the nested if.

**Review by zofskeez — COMMENTED** [review]
:clap:

File: ui/app/models/database/role.js
Comment (hellobontempo) [reply ui/app/models/database/role.js:15]: Yeah - I agree. Going to leave as is (maybe this was also decided during development and the designs weren't updated)

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js:33]: Thank you for catching! I hadn't updated the documentation

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js:150]: Great point! I wonder why it was separated in the old component 🤔

File: ui/lib/core/addon/components/search-select.hbs
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.hbs:19]: Yeah - this is the block I was referring to in the dev channel last week. We only ever use `InputSearch` or `StringList` so I wondered about directly placing those here. I think yield would make sense in a figure iteration, especially if we move the management of selections to the parent. 

What do you think @zofskeez?

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js:150]: Oh! I think it was separated because if there are no models and no options, we want to return completely from `fetchOptions` - which wouldn't happen if this was combined to `if (!this.args.models && this.args.options)`

But there may be a cleaner way to relay this that doesn't require a nested `if` statement - by adding a return before the for loop below? What do you think?

Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js:150]:
```suggestion
    if (!this.args.models && !this.args.options) {
        const options = this.args.options;
        if (options.some((e) => Object.keys(e).includes('groupName'))) {
          this.dropdownOptions = options;
        } else {
          this.dropdownOptions = [...this.addSearchText(options)];
          this.selectedOptions = this.formatSelectedAndUpdateDropdown(this.selectedOptions);
        }
  }
  
  if (!this.args.models) return
```

Comment (zofskeez) [reply ui/lib/core/addon/components/search-select.js:150]: > Oh! I think it was separated because if there are no models and no options, we want to return completely from `fetchOptions` - which wouldn't happen if this was combined to `if (!this.args.models && this.args.options)`
> 
> But there may be a cleaner way to relay this that doesn't require a nested `if` statement - by adding a return before the for loop below? What do you think @zofskeez ?

Oh ya good point I missed that! I think it's probably better to leave as is. It was a nit on my part and not worth risky a potential regression.

File: ui/lib/core/addon/components/search-select.hbs
Comment (zofskeez) [reply ui/lib/core/addon/components/search-select.hbs:19]: I think it's good to leave as is so we don't blow out the scope.

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [inline ui/lib/core/addon/components/search-select.js:139]: I confess I kind of lost track of `this.options` at this point and am not confident I understand what this `if` block is checking for...that nothing has been set at all? 

The new value of `this.dropdownOptions` will always be truthy, since I 'reset' it to an empty array at the beginning of this method.

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js]: Discussed with Ivana on slack and we decided to hold off making this change for now

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hashishaw) [inline ui/app/templates/components/get-credentials-card.hbs:29]: A `button` within a form automatically has `type="submit"` so this can be removed!

File: ui/app/templates/components/input-search.hbs
Comment (hashishaw) [inline ui/app/templates/components/input-search.hbs:1]: nice

File: ui/tests/integration/components/get-credentials-card-test.js
Comment (hashishaw) [inline ui/tests/integration/components/get-credentials-card-test.js:54]: Is `@placeholder` required now, is that why we added it here?

File: ui/lib/core/addon/components/form-field.hbs
Comment (hashishaw) [inline ui/lib/core/addon/components/form-field.hbs:88]: I wonder if we should instead use `isSectionHeader` and default to false, especially since it seems like using it as section header is an old and unpopular approach at this point

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hashishaw) [reply ui/app/templates/components/get-credentials-card.hbs:15]: I think we should keep the `403` functionality. The use case there is that if someone is filling a form but doesn't have the permissions to see the list, they can still fill out the form correctly as long as they know the name of the item they need for that input. 
That being said, if it's being set by the parent it doesn't really make sense to force the child (SearchSelect) to render, since the parent already knows we don't really want the SearchSelect, so the pattern you have here makes sense to extend. All in all, I think `shouldUseFallback` in SearchSelect should be a component attribute rather than a passed attribute, and we'll leave it to the parents who know they want to use the fallback to just render the desired fallback instead

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hashishaw) [reply ui/app/templates/components/get-credentials-card.hbs:26]: The functionality is essentially the same so I don't think it needs to be backported

**Review by hashishaw — CHANGES_REQUESTED** [review]
Overall looks good! I think the biggest improvement would be to make `shouldUseFallback` an internally tracked attribute within `SearchSelect` so that it's really only in charge of handling the case where the model fetch gets a 403.

**Review by hashishaw — APPROVED** [review]
Other than the "not" attribute LGTM!

File: ui/app/templates/components/get-credentials-card.hbs
Comment (hellobontempo) [reply ui/app/templates/components/get-credentials-card.hbs:29]: Prettier gets angry if the button doesn't have `type` attribute, i can change it to `type="button"` ?
Screen Shot 2022-09-29 at 10 57 55 AM

File: ui/lib/core/addon/components/form-field.hbs
Comment (hellobontempo) [reply ui/lib/core/addon/components/form-field.hbs:88]: I wondered the same! I was just nervous to change the default, but I can dig around and find all the instances it's intentionally used _as_ a header - shouldn't be too hard, I just have to track down all model attrs with `editType: 'search-select'`

File: ui/lib/core/addon/components/search-select.js
Comment (hashishaw) [inline ui/lib/core/addon/components/search-select.js:16]: nit: Can we add `@` to the attributes in this example so it's true to life?

File: ui/lib/core/addon/components/search-select.js
Comment (hashishaw) [inline ui/lib/core/addon/components/search-select.js:45]: Not sure we should tie the label so closely to the item noun as it's used in the create new option. More and more of our labels are descriptive or instructive rather than just what the thing is, so maybe we use this by default but provide an optional arg where it can be overridden. Also, since label is optional it would be good to set a default for the item noun in `addCreateOption`, (like "item")

File: ui/lib/core/addon/components/search-select.hbs
Comment (hellobontempo) [inline ui/lib/core/addon/components/search-select.hbs:32]: was previously 
```
 {{! template-lint-configure simple-unless "warn" }}
  {{#unless (gte this.selectedOptions.length this.selectLimit)}}

File: ui/lib/core/addon/components/search-select.js
Comment (hellobontempo) [reply ui/lib/core/addon/components/search-select.js]: good call!

**Pull request merged**
